Documentation/RepoApi.md

```diff
@@ -0,0 +1,137 @@
+
+This document describes the basic structure of this "Repo API", and the requirements imposed on a repo that needs
+to participate in source-build.
+
+## build.cmd/sh
+
+Each repo will have two files that implement the repo API:
+
+* build.cmd
+* build.sh
+
+The `.cmd` file will be called when running on Windows, while the `.sh` file will be called on all other operating systems.
+
+These scripts will contain the logic necessary in order to build, test, publish the repo's assets.
+
+source-build will invoke these scripts on each repo passing in:
+
+1. An action to perform
+1. A set of parameters
+
+Individual repos are free to implement more features in these build scripts than what is required by source-build. For
+example, a repo may want to provide a default action. These extra behaviors are not defined by source-build.
+
+### Actions
+
+To start, there are a number of actions source-build will need the repos to implement.  The actions are:
+
+1. `Restore` - Sync any external dependencies/tools necessary to build the repo.
+1. `Build` - Assemble the product from the current source contained in the repo.
+1. `Publish` - Put the assembled product in the specified location.
+1. `Test` - Ensure the assembled product works/acts correctly.
+
+This list of actions can be added to in the future as more scenarios are implemented.
+
+source-build formats arguments using a style that is compatible with MSBuild. The reasoning for this is that the
+majority of repos have an MSBuild based build system. Using this format will allow for arguments to easily be passed
+through to MSBuild. Similarly, both action and parameter arguments are case-insensitive, just like MSBuild default
+behavior.
+
+Actions will take the form of `-t:ActionName`, just like MSBuild targets are passed to `msbuild.exe`.
+
+### Parameters
+
+Each action above will take a set of parameters used to pass information into the action and control its behavior.
+
+Parameters will take the form of `-p:ParameterName=ParameterValue`, just like MSBuild properties are passed to `msbuild.exe`.
+One slight variation here is the usage of `-p` instead of `/p`.  MSBuild supports both forms, but source-build chooses `-p` because
+it meets [POSIX recommendations](https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html) for command line parameters
+and avoids any confusion with file paths on Unix machines.
+
+In order to avoid conflicts between existing MSBuild properties used by the repos and the properties passed in by source-build,
+source-build will append a prefix to every property it passes in: `DotNet`.
+
+#### Detecting build from source
+
+The first parameter used by source-build will be a value indicating whether the build is being performed in "source only" mode. This
+means repos cannot use tools/packages/references that haven't already been built by the current build.
```

Comment:
I wonder if we need a way to enforce this? Perhaps we enforce it by limiting where it can restore stuff in the “restore” command.

Reply:
If we go down that route, we need to enforce downloading of non-nuget artifacts as well

Reply:
I need to rephrase this, because this parameter actually is turning out to mean something different.

I'm not sure exactly how enforcing would work.  Ideas are welcome.